packages/server/src/modules/user/seed-users-with-books.command.ts

```diff
@@ -0,0 +1,518 @@
+                  requests: true,
+                  reservations: true,
+                },
+              },
+              sales: true,
+            },
+          },
+        },
+      });
+
+      // Connect requests and reservations to the sales
+      let i = 0;
+      for (const sale of sales) {
+        await this.#updateReservationsRequestsAndCart(sale, cart);
+
+        if (i % 3 === 2) {
+          await this.#refundBookCopy(sale, retailLocationId);
+        }
+        i++;
+      }
+    } catch (error) {
+      this.logger.error("Something went wrong: ", error);
+    }
+  }
+
+  async #updateReservationsRequestsAndCart(
+    sale: {
+      bookCopy: {
+        book: {
+          reservations: Reservation[];
+          requests: BookRequest[];
+        } & Book;
+      };
+    } & Sale,
+    cart: Cart,
+  ) {
+    const book = sale.bookCopy.book;
+    await this.prisma.cartItem.delete({
+      where: {
+        cartId_bookId: {
+          bookId: book.id,
+          cartId: cart.id,
+        },
+      },
+    });
+
+    if (book.requests.length > 0) {
+      const request = book.requests.find(
+        ({ userId, saleId }) =>
+          userId === sale.purchasedById && saleId === null,
+      );
+      if (!request) {
+        throw new Error(
+          `Could not update the request associated with this book: ${book.id}`,
+        );
+      }
+
+      await this.prisma.bookRequest.update({
+        where: { id: request.id },
+        data: {
+          saleId: sale.id,
+        },
+      });
+    }
+
+    if (book.reservations.length > 0) {
+      const reservation = book.reservations.find(
+        ({ userId, saleId }) =>
+          userId === sale.purchasedById && saleId === null,
+      );
+      if (!reservation) {
+        throw new Error(
+          `Could not update the request associated with this book: ${book.id}`,
+        );
+      }
+
+      await this.prisma.reservation.update({
+        where: { id: reservation.id },
+        data: {
+          saleId: sale.id,
+        },
+      });
+    }
+  }
+
+  async #refundBookCopy(
+    { id, bookCopyId, bookCopy }: Sale & { bookCopy: BookCopy },
+    retailLocationId: string,
+  ) {
+    const [newBookCopyCode] = await this.bookService.calculateBookCodes(
+      this.prisma,
+      [bookCopy.bookId],
+      retailLocationId,
+      true,
+    );
+
+    return this.prisma.bookCopy.update({
+      where: {
+        id: bookCopyId,
+      },
+      data: {
+        code: newBookCopyCode,
+        // Original code must be updated only the first time a book is returned, so when there is no original code set.
+        ...(bookCopy.originalCode ? {} : { originalCode: bookCopy.code }),
+        updatedAt: new Date(),
+        sales: {
+          update: {
+            where: {
+              id,
+            },
+            data: {
+              refundedAt: new Date(),
+            },
+          },
+        },
+      },
+    });
+  }
+
+  async #clearRetailLocationSeededUsers() {
```

Comment:
Uhm question here: are you sure you cannot leverage the cascade effect here? I remember that Yusuf inside reset year function was able to leverage `ONDELETE CASCADE` set on the DB level. Maybe not all of these checks and operations are really necessary.

Reply:
That part of the code is directly taken from the `book-copy.resolver`, and there Yusuf made it that way so I trusted him on it; I remember I already tried running the code without this bit and it didn't update correctly at the time but I'll double check.